src/kmeans.jl

```diff
@@ -64,17 +63,18 @@ end
 
 # core k-means skeleton
 function _kmeans!(
-    x::Matrix{T},                   # in: sample matrix (d x n)
-    w::Union{Nothing, Vector{T}},      # in: sample weights (n)
-    centers::Matrix{T},             # in/out: matrix of centers (d x k)
-    assignments::Vector{Int},       # out: vector of assignments (n)
-    costs::Vector{T},               # out: costs of the resultant assignments (n)
-    counts::Vector{Int},            # out: the number of samples assigned to each cluster (k)
-    cweights::Vector{Float64},      # out: the weights of each cluster
-    maxiter::Int,                   # in: maximum number of iterations
-    tol::Real,                      # in: tolerance of change at convergence
-    displevel::Int,                 # in: the level of display
-    distance::SemiMetric) where T<:AbstractFloat             # in: function to calculate the distance with
+    x::AbstractMatrix{T},         # in: sample matrix (d x n)
+    w::Union{Nothing, Vector{T}}, # in: sample weights (n)
```

Comment:
It would also be nice to change `Vector{T}` (and `Vector{Int}` etc) into `AbstractVector{T}` (resp. `AbstractVector{Int}`)  here and in the other functions.

Reply:
Ok thanks @alyst , I'll try to apply the changes in a way that's easily reviewable and if it looks like it might be too much then maybe another PR would be better.

Reply:
Actually I'm not sure that makes a lot of sense given that the caller function allocates those vectors explicitly?